diff --git a/src/ancestral.c b/src/ancestral.c
index a8359c7..9cd2754 100644
--- a/src/ancestral.c
+++ b/src/ancestral.c
@@ -12,7 +12,7 @@ the GNU public licence. See http://www.opensource.org for details.
 
 #include "ancestral.h"
 
-void Sample_Ancestral_Seq(int fullmutmap, int fromprior, t_tree *tree)
+void Sample_Ancestral_Seq(MTRand* rand, int fullmutmap, int fromprior, t_tree *tree)
 {
   int rate_cat;
   int i,j,k,l;
@@ -28,7 +28,7 @@ void Sample_Ancestral_Seq(int fullmutmap, int fromprior, t_tree *tree)
   
   if(tree->is_mixt_tree == YES)
     {
-      MIXT_Sample_Ancestral_Seq(fullmutmap,fromprior,tree);
+      MIXT_Sample_Ancestral_Seq(rand,fullmutmap,fromprior,tree);
       return;
     }
 
@@ -79,18 +79,20 @@ void Sample_Ancestral_Seq(int fullmutmap, int fromprior, t_tree *tree)
       for(j=0;j<tree->mod->ras->n_catg;j++) sum += probs[j];
       for(j=0;j<tree->mod->ras->n_catg;j++) probs[j]/=sum;
 
-      rate_cat = Sample_i_With_Proba_pi(probs,tree->mod->ras->n_catg);
+      rate_cat = Sample_i_With_Proba_pi(rand,probs,tree->mod->ras->n_catg);
            
       n_mut = 0;
       if(tree->n_root != NULL)
         {          
-          Sample_Ancestral_Seq_Pre(tree->n_root,tree->n_root->v[1],tree->n_root->b[1],
+          Sample_Ancestral_Seq_Pre(rand,
+                                   tree->n_root,tree->n_root->v[1],tree->n_root->b[1],
                                    i,rate_cat,
                                    muttype,muttime,muttax,&n_mut,
                                    fullmutmap,fromprior,tree);
 
           
-          Sample_Ancestral_Seq_Pre(tree->n_root,tree->n_root->v[2],tree->n_root->b[2],
+          Sample_Ancestral_Seq_Pre(rand,
+                                   tree->n_root,tree->n_root->v[2],tree->n_root->b[2],
                                    i,rate_cat,
                                    muttype,muttime,muttax,&n_mut,
                                    fullmutmap,fromprior,tree);
@@ -99,7 +101,8 @@ void Sample_Ancestral_Seq(int fullmutmap, int fromprior, t_tree *tree)
         }
       else
         {
-          Sample_Ancestral_Seq_Pre(tree->a_nodes[0],tree->a_nodes[0]->v[0],tree->a_nodes[0]->b[0],
+          Sample_Ancestral_Seq_Pre(rand,
+                                   tree->a_nodes[0],tree->a_nodes[0]->v[0],tree->a_nodes[0]->b[0],
                                    i,rate_cat,
                                    muttype,muttime,muttax,&n_mut,
                                    fullmutmap,fromprior,tree);
@@ -160,7 +163,8 @@ void Sample_Ancestral_Seq(int fullmutmap, int fromprior, t_tree *tree)
 //////////////////////////////////////////////////////////////
 //////////////////////////////////////////////////////////////
 
-void Sample_Ancestral_Seq_Pre(t_node *a, t_node *d, t_edge *b,
+void Sample_Ancestral_Seq_Pre(MTRand* rand,
+                              t_node *a, t_node *d, t_edge *b,
                               int site, int r_cat,
                               int *muttype, phydbl *muttime, int *muttax, int *n_mut,
                               int fullmutmap, int fromprior, t_tree *tree)
@@ -186,11 +190,11 @@ void Sample_Ancestral_Seq_Pre(t_node *a, t_node *d, t_edge *b,
       sum = 0.0;
       for(i=0;i<ns;++i) sum += probs[i];
       for(i=0;i<ns;++i) probs[i] /= sum;
-      sa = Sample_i_With_Proba_pi(probs,ns);
+      sa = Sample_i_With_Proba_pi(rand,probs,ns);
     }
   else if(a == tree->n_root)
     {
-      sa = Sample_Ancestral_Seq_Core(NULL,tree->n_root,NULL,r_cat,site,tree);
+      sa = Sample_Ancestral_Seq_Core(rand,NULL,tree->n_root,NULL,r_cat,site,tree);
     }
   else
     {
@@ -199,9 +203,9 @@ void Sample_Ancestral_Seq_Pre(t_node *a, t_node *d, t_edge *b,
                         tree->mod->io->state_len);
     }
 
-  sd = Sample_Ancestral_Seq_Core(a,d,b,r_cat,site,tree);
+  sd = Sample_Ancestral_Seq_Core(rand,a,d,b,r_cat,site,tree);
   
-  if(fullmutmap == YES) Map_Mutations(a,d,sa,sd,b,site,r_cat,muttype,muttime,muttax,n_mut,tree);
+  if(fullmutmap == YES) Map_Mutations(rand,a,d,sa,sd,b,site,r_cat,muttype,muttime,muttax,n_mut,tree);
     
   if(d->tax) return;
   else
@@ -210,7 +214,7 @@ void Sample_Ancestral_Seq_Pre(t_node *a, t_node *d, t_edge *b,
         {
           if(d->v[i] != a && d->b[i] != tree->e_root)
             {
-              Sample_Ancestral_Seq_Pre(d,d->v[i],d->b[i],site,r_cat,muttype,muttime,muttax,n_mut,fullmutmap,fromprior,tree);
+              Sample_Ancestral_Seq_Pre(rand,d,d->v[i],d->b[i],site,r_cat,muttype,muttime,muttax,n_mut,fullmutmap,fromprior,tree);
             }
         }
     }
@@ -219,7 +223,7 @@ void Sample_Ancestral_Seq_Pre(t_node *a, t_node *d, t_edge *b,
 //////////////////////////////////////////////////////////////
 //////////////////////////////////////////////////////////////
 
-int Sample_Ancestral_Seq_Core(t_node *a, t_node *d, t_edge *b, int r_cat, int site, t_tree *tree)
+int Sample_Ancestral_Seq_Core(MTRand* rand, t_node *a, t_node *d, t_edge *b, int r_cat, int site, t_tree *tree)
 {
   int i,j;
   int ns;
@@ -248,7 +252,7 @@ int Sample_Ancestral_Seq_Core(t_node *a, t_node *d, t_edge *b, int r_cat, int si
       sum = 0.0;
       for(i=0;i<ns;++i) sum += probs[i];
       for(i=0;i<ns;++i) probs[i] /= sum;
-      state = Sample_i_With_Proba_pi(probs,ns);
+      state = Sample_i_With_Proba_pi(rand,probs,ns);
     }
   else
     {
@@ -299,7 +303,7 @@ int Sample_Ancestral_Seq_Core(t_node *a, t_node *d, t_edge *b, int r_cat, int si
           for(i=0;i<ns;++i) sum += probs[i];
           for(i=0;i<ns;++i) probs[i] /= sum;
           
-          state = Sample_i_With_Proba_pi(probs,ns);
+          state = Sample_i_With_Proba_pi(rand,probs,ns);
           
           d->c_seq_anc->state[site] = Reciproc_Assign_State(state,tree->io->datatype);
         }
@@ -327,7 +331,7 @@ int Sample_Ancestral_Seq_Core(t_node *a, t_node *d, t_edge *b, int r_cat, int si
           for(i=0;i<ns;++i) sum += probs[i];
           for(i=0;i<ns;++i) probs[i] /= sum;
           
-          state = Sample_i_With_Proba_pi(probs,ns);
+          state = Sample_i_With_Proba_pi(rand,probs,ns);
           
           d->c_seq_anc->state[site] = Reciproc_Assign_State(state,tree->io->datatype);
         }      
@@ -342,7 +346,7 @@ int Sample_Ancestral_Seq_Core(t_node *a, t_node *d, t_edge *b, int r_cat, int si
 //////////////////////////////////////////////////////////////
 //////////////////////////////////////////////////////////////
 
-void Map_Mutations(t_node *a, t_node *d, int sa, int sd, t_edge *b, int site, int rcat, int *muttype, phydbl *muttime, int *muttax, int *n_mut, t_tree *tree)
+void Map_Mutations(MTRand* rand, t_node *a, t_node *d, int sa, int sd, t_edge *b, int site, int rcat, int *muttype, phydbl *muttime, int *muttax, int *n_mut, t_tree *tree)
 {
   int i,j;
   phydbl *probs,*all_probs;
@@ -416,13 +420,13 @@ void Map_Mutations(t_node *a, t_node *d, int sa, int sd, t_edge *b, int site, in
         {
           // Sample a time for the first mutation conditional on at least one mutation
           // occurring (see formula 2.1 in Hobolth and Stone, 2009).
-          u = Uni();
+          u = Uni(rand);
           tlast = -log(1. - u*(1.-exp(Q[sa*ns+sa]*T)))/-Q[sa*ns+sa];
         }
       else
         {
           // Sample a time for the next mutation
-          tlast = tlast + Rexp(-Q[slast*ns+slast]);
+          tlast = tlast + Rexp(rand,-Q[slast*ns+slast]);
         }
       
       // Select the appropriate vector of change probabilities
@@ -445,7 +449,7 @@ void Map_Mutations(t_node *a, t_node *d, int sa, int sd, t_edge *b, int site, in
                          tree->rates ? tree->rates->cur_l[d->num] : b->l->v,
                          n_mut_branch);
           
-          snew = Sample_i_With_Proba_pi(probs,ns);
+          snew = Sample_i_With_Proba_pi(rand,probs,ns);
           
           // Record mutation type
           mut[slast*ns+snew]++;
